### PR TITLE
InstCombine: Fold shufflevector(select) and shufflevector(phi)

### PR DESCRIPTION
**Pull request by MatzeB**
- Transform `shufflevector(select(c, x, y), C)` to
  `select(c, shufflevector(x, C), shufflevector(y, C))` by re-using
  the `FoldOpIntoSelect` helper.
- Transform `shufflevector(phi(x, y), C)` to
  `phi(shufflevector(x, C), shufflevector(y, C))` by re-using the
  `foldOpInotPhi` helper.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp [hunk 1]
```diff
@@ -2900,6 +2900,17 @@ Instruction *InstCombinerImpl::visitShuffleVectorInst(ShuffleVectorInst &SVI) {
   if (Instruction *I = foldIdentityPaddedShuffles(SVI))
     return I;
 
+  if (Constant *C = dyn_cast<Constant>(RHS)) {
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:2903]:
```suggestion
  if (isa<Constant>(RHS)) {
```

**Review by dtcxzyw — APPROVED** [review]
LGTM. Nice catch!

**Review by goldsteinn — APPROVED** [review]
lgtm

File: llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp [hunk 2]
```diff
@@ -2900,6 +2900,17 @@ Instruction *InstCombinerImpl::visitShuffleVectorInst(ShuffleVectorInst &SVI) {
...
+    if (SelectInst *SI = dyn_cast<SelectInst>(LHS)) {
```
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:2904]: (style) use auto *SI

File: llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp [hunk 3]
```diff
@@ -2900,6 +2900,17 @@ Instruction *InstCombinerImpl::visitShuffleVectorInst(ShuffleVectorInst &SVI) {
...
+      if (Instruction *I = FoldOpIntoSelect(SVI, SI))
+        return I;
+    }
+    if (PHINode *PN = dyn_cast<PHINode>(LHS)) {
```
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:2908]: (style) use auto *PN

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

File: llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp (hunk 1 above)
Comment (MatzeB) [reply llvm/lib/Transforms/InstCombine/InstCombineVectorOps.cpp:2903]: Good point! Will go with `match(RHS, M_Constant())` instead as that seems to match the style of the surrounding code pattern.

**Pull request merged**